pkg/kubelet/collector/interfaces.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collector
+
+type Interface interface {
+	// Starts the collector process if necessary, e.g., the builtin cadvisor
+	// Otherwise, does nothing and just returns
+	Start() error
+
+	// Returns the basic machine spec
+	// Currently used to report to k8s master the basic information about a node
+	MachineInfo() (*MachineInfo, error)
+
+	// Returns the software version of various components of a node, e.g., kernel, container runtime, etc.
+	// Currently used to report back to k8s master the versions of relevant software on a node
+	VersionInfo() (*VersionInfo, error)
+
+	// Returns the file system information of a given label or mount point
+	// Currently used by kubelet to check if there's enough disk space to schedule new pods and
+	// by the image manager to decide if garbage collection should start
+	FsInfo(fslabel string) (*FsInfo, error)
+
+	// Returns a channel for watching requested event types
+	// Currently used by oom watcher to detect oom events.
+	WatchEvents(request *Request) (chan *Event, error)
+
+	// Get detailed metrics information (cpu, memory, disk, network, etc.) about one or more containers
+	// When isRawContainer=true, containerName refers to a raw Linux container; otherwise it refers to a Docker container
+	// When isRawContainer=true and subcontainer=true, returns not only the metrics information of this container, but
+	// also the child containers
+	// Returned value is a map of container name to its metrics data, and this data is in a collector-specific format
```

Comment:
What is meant by `container name` here? Kuberenetes has a concept of a container name, but it's scoped to a pod and not guaranteed to be globally unique. I think it would be better to create a `ContainerRef` struct, or use a [`ContainerId`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/runtime.go#L147) (and define a `raw` type)